pkg/trtl/replica/sync.go

```diff
@@ -0,0 +1,491 @@
+			// NOTE: this is not fully safe, since there is no logical barrier
+			// between this check and the send on the chanel, however the sender
+			// keeps consuming messages to prevent race conditions.
+			if ok := sender.Send(msg); !ok {
+				// NOTE: breaking objects loop not the namespaces loop or returning to
+				// ensure that all iters get released. This may cause some additional
+				// Send calls to happen, but the sender will simply ignore them.
+				break objects
+			}
+			nVersions++
+		}
+
+		if err = iter.Error(); err != nil {
+			log.Error().Err(err).Str("namespace", namespace).Msg("could not iterate over namespace")
+		}
+
+		// Ensure the iterator is released, note even if break objects occurs, the
+		// iterator should be released at this line of code since there is no return.
+		iter.Release()
+		nNamespaces++
+	}
+
+	// Send a sync complete message to let the remote know that the pull phase is
+	// complete and they can start the push phase.
+	sender.Send(&replica.Sync{Status: replica.Sync_COMPLETE})
+	log.Trace().Msg("initiator phase 1 complete")
+
+	log.Debug().
+		Uint64("versions", nVersions).
+		Uint64("namespaces", nNamespaces).
+		Msg("version vectors sent to remote peer")
+}
+
+// initiatorPhase2 starts right after initiatorPhase1 and runs as long as messages will
+// be coming from the remote. This is the only initiator go routine that will receive
+// messages, so an intermediate read routine is not necessary. This phase is also
+// responsible for closing the sender go routine because the phase is finished when it
+// gets a COMPLETE message from the remote. This phase handles incoming messages from
+// the remote by responding to CHECK requests sending later versions to the remote (but
+// ignoring if local versions are equal or earlier), and handling REPAIR and error.
+func (r *Service) initiatorPhase2(ctx context.Context, wg *sync.WaitGroup, log zerolog.Logger, sender *streamSender, stream gossipStream) {
+	// Ensure that this routine signals when it exits
+	defer wg.Done()
+	log.Trace().Msg("starting initiator phase 2")
+
+	// Ensure that we close the sending channel when this routine exits to prevent
+	// deadlocks if this phase ends prematurely (e.g. the timeout expires).
+	defer sender.Close()
+
+	// Track how many check versions, updates, and repairs we get in phase 2
+	var versions, updates, repairs uint64
+	var err error
+
+gossip:
+	for {
+		// Check to make sure the deadline isn't over
+		select {
+		case <-ctx.Done():
+			log.Debug().Msg("context canceled while trying to recv messages from remote")
+			return
+		default:
+		}
+
+		// Read the next message from the remote replica
+		var sync *replica.Sync
+		if sync, err = stream.Recv(); err != nil {
+			if err != io.EOF {
+				// If the error is not EOF then something has gone wrong.
+				log.Error().Err(err).Msg("anti-entropy aborted early with recv error")
+			}
+			return
+		}
+
+		switch sync.Status {
+		case replica.Sync_CHECK:
```

Comment:
Is this the mirror CHECK mentioned in Gossip?

Reply:
Yes!